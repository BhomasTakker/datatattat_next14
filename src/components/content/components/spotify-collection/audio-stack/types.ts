import { ContainerHeightOptions } from "@/components/page/components/stack/types";
import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
	height: ContainerHeightOptions;
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
