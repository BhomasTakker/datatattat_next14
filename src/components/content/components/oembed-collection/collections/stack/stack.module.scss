.root {
	.list {
		display: flex;
		flex-direction: column;
		align-items: center;
		width: 100%;
		li {
			// sm, md, etc
			width: 100%;
		}
	}
}
