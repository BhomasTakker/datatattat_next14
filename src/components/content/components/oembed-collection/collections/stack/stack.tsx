import { OEmbed } from "@/types/data-structures/oembed";
import { ClientOembed } from "../../../content-oembed/client-component";
import styles from "./stack.module.scss";

export type OembedStackProps = {};

const renderOembed = (item: OEmbed) => {
	const { src, html } = item || {};

	return (
		<li key={item.id} data-testid="content-oembed" className={styles.root}>
			{html && <ClientOembed html={html} />}
		</li>
	);
};

const renderMethod = (collection: OEmbed[] = [], props: OembedStackProps) => {
	return (
		<ul className={`${styles.list}`}>
			{collection.map((item) => renderOembed(item))}
		</ul>
	);
};

const oembedStack = {
	styles,
	renderMethod,
};

export default oembedStack;
