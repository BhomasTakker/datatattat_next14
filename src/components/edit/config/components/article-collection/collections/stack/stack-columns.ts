import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const STACK_COLUMNS_CONFIG: InputListProps = {
	id: "stackColumns",
	type: EditInputs.inputList,
	label: "Stack Columns",
	inputs: [],
};
