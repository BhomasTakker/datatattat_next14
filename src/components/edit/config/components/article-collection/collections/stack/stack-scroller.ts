import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const STACK_SCROLLER_CONFIG: InputListProps = {
	id: "stackScroller",
	type: EditInputs.inputList,
	label: "Stack Scroller",
	inputs: [],
};
