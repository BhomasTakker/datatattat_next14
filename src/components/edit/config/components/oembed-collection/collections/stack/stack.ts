import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const OEMBED_STACK_CONFIG: InputListProps = {
	id: "stack",
	type: EditInputs.inputList,
	label: "Stack",
	inputs: [
		{
			id: "oembedComponentTitle",
			type: EditInputs.title,
			title: "Oembed Stack Component",
		},
	],
};
