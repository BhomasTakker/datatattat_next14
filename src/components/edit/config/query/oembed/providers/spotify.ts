import { EditInputs } from "@/components/edit/inputs/inputs";
import { GenericInput, InputListProps } from "@/types/edit/inputs/inputs";

const spotifyOembedInputs: GenericInput[] = [
	{
		id: "asset",
		type: EditInputs.select,
		label: "Spotify Asset",
		defaultValue: "episode",
		options: ["album", "artist", "show", "episode", "playlist", "track"],
	},
	{
		id: "contentId",
		type: EditInputs.text,
		label: "Content ID",
		validation: {
			required: true,
			minLength: 22,
			maxLength: 22, // Spotify content IDs are typically 22 characters long
		},
	},
];

export const SPOTIFY_OEMBED_CONFIG: GenericInput = {
	id: "spotifyOembedParams",
	type: EditInputs.inputList,
	label: "Spotify Oembed Parameters",
	createObject: false,

	inputs: spotifyOembedInputs,
};

const params: InputListProps = {
	id: "params",
	type: EditInputs.inputList,
	label: "SPOTIFY OEMBED QUERY",

	inputs: spotifyOembedInputs,
};

export const SPOTIFY_COLLECTION_OEMBED_CONFIG: InputListProps = {
	id: "spotifyOembedQuery",
	type: EditInputs.inputList,
	label: "SPOTIFY OEMBED QUERY",
	inputs: [
		{
			id: "parameters",
			type: EditInputs.title,
			title: "Oembed Query Parameters",
		},
		params,
	],
};
