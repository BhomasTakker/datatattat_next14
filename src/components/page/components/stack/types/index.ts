import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
