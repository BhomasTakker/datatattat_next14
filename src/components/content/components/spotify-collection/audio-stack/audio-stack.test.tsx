import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import spotifyAudioStack from "./audio-stack";
import { ContainerHeightOptions } from "@/components/page/components/stack/types";
import { SearchTypes } from "@/types/api/spotify";
import { SpotifyCollectionItem } from "./types";

// Mock the client component since it uses hooks and API calls
jest.mock("./audio-stack-client", () => {
	return {
		__esModule: true,
		AudioStackClientComponent: ({ items, height }: any) => (
			<div data-testid="audio-stack-client">
				<div data-testid="height-option">{height}</div>
				<div data-testid="items-count">{items.length}</div>
				{items.map((item: SpotifyCollectionItem) => (
					<div key={item.id} data-testid={`item-${item.id}`}>
						{item.title}
					</div>
				))}
			</div>
		),
	};
});

const mockItems: SpotifyCollectionItem[] = [
	{
		id: "test-track-1",
		title: "Test Song 1",
		src: "https://spotify.com/track1",
		description: "Test Description 1",
		guid: "guid-track-1",
		variant: "spotify-track",
		media: {
			type: SearchTypes.Track,
		},
	},
	{
		id: "test-track-2",
		title: "Test Song 2",
		src: "https://spotify.com/track2",
		description: "Test Description 2",
		guid: "guid-track-2",
		variant: "spotify-album",
		media: {
			type: SearchTypes.Album,
		},
	},
];

describe("Audio Stack Component Test Suite", () => {
	describe("spotifyAudioStack object", () => {
		it("should have styles property", () => {
			expect(spotifyAudioStack.styles).toBeDefined();
		});

		it("should have renderMethod property", () => {
			expect(spotifyAudioStack.renderMethod).toBeDefined();
			expect(typeof spotifyAudioStack.renderMethod).toBe("function");
		});
	});

	describe("renderMethod", () => {
		it("should render AudioStackClientComponent with correct props", () => {
			const options = { height: ContainerHeightOptions.MD };
			const result = spotifyAudioStack.renderMethod(mockItems, options);

			render(result);

			expect(screen.getByTestId("audio-stack-client")).toBeInTheDocument();
			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.MD
			);
			expect(screen.getByTestId("items-count")).toHaveTextContent("2");
		});

		it("should render with empty items array when no items provided", () => {
			const options = { height: ContainerHeightOptions.SM };
			const result = spotifyAudioStack.renderMethod(undefined, options);

			render(result);

			expect(screen.getByTestId("audio-stack-client")).toBeInTheDocument();
			expect(screen.getByTestId("items-count")).toHaveTextContent("0");
		});

		it("should render with different height options", () => {
			const options = { height: ContainerHeightOptions.LG };
			const result = spotifyAudioStack.renderMethod(mockItems, options);

			render(result);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.LG
			);
		});

		it("should render all provided items", () => {
			const options = { height: ContainerHeightOptions.MD };
			const result = spotifyAudioStack.renderMethod(mockItems, options);

			render(result);

			expect(screen.getByTestId("item-test-track-1")).toHaveTextContent(
				"Test Song 1"
			);
			expect(screen.getByTestId("item-test-track-2")).toHaveTextContent(
				"Test Song 2"
			);
		});

		it("should handle single item", () => {
			const singleItem = [mockItems[0]];
			const options = { height: ContainerHeightOptions.MD };
			const result = spotifyAudioStack.renderMethod(singleItem, options);

			render(result);

			expect(screen.getByTestId("items-count")).toHaveTextContent("1");
			expect(screen.getByTestId("item-test-track-1")).toBeInTheDocument();
			expect(screen.queryByTestId("item-test-track-2")).not.toBeInTheDocument();
		});

		it("renders spotifyAudioStack unchanged", () => {
			const options = { height: ContainerHeightOptions.MD };
			const result = spotifyAudioStack.renderMethod(mockItems, options);
			const { container } = render(result);

			expect(container).toMatchSnapshot();
		});
	});

	describe("AudioStackProps type validation", () => {
		it("should work with ContainerHeightOptions.XS", () => {
			const options = { height: ContainerHeightOptions.XS };
			const result = spotifyAudioStack.renderMethod(mockItems, options);
			render(result);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.XS
			);
		});

		it("should work with ContainerHeightOptions.SM", () => {
			const options = { height: ContainerHeightOptions.SM };
			const result = spotifyAudioStack.renderMethod(mockItems, options);
			render(result);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.SM
			);
		});

		it("should work with ContainerHeightOptions.MD", () => {
			const options = { height: ContainerHeightOptions.MD };
			const result = spotifyAudioStack.renderMethod(mockItems, options);
			render(result);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.MD
			);
		});

		it("should work with ContainerHeightOptions.LG", () => {
			const options = { height: ContainerHeightOptions.LG };
			const result = spotifyAudioStack.renderMethod(mockItems, options);
			render(result);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.LG
			);
		});

		it("should work with ContainerHeightOptions.XL", () => {
			const options = { height: ContainerHeightOptions.XL };
			const result = spotifyAudioStack.renderMethod(mockItems, options);
			render(result);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.XL
			);
		});
	});
});
