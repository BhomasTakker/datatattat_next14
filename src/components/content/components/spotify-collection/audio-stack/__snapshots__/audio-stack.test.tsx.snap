// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Audio Stack Component Test Suite renderMethod renders spotifyAudioStack unchanged 1`] = `
<div>
  <div
    data-testid="audio-stack-client"
  >
    <div
      data-testid="height-option"
    >
      MD
    </div>
    <div
      data-testid="items-count"
    >
      2
    </div>
    <div
      data-testid="item-test-track-1"
    >
      Test Song 1
    </div>
    <div
      data-testid="item-test-track-2"
    >
      Test Song 2
    </div>
  </div>
</div>
`;
